samples/swift/uidemo/ChatViewController.swift

```diff
@@ -0,0 +1,251 @@
+      // we don't set it before trying to populate our view our app will crash.
+      self.collectionView.dataSource = self.collectionViewDataSource
+      
+      self.collectionViewDataSource.populateCellWithBlock { (anyCell, data) in
+        guard let cell = anyCell as? ChatCollectionViewCell else {
+          fatalError("Unexpected collection view cell class \(anyCell.self)")
+        }
+        
+        let chat = Chat(snapshot: data as! FIRDataSnapshot)!
+        
+        cell.textLabel.text = chat.text
+        
+        let leftRightPadding: CGFloat = 24
+        let rect = ChatCollectionViewCell.boundingRectForText(cell.textLabel.text ?? "",
+          maxWidth: self.view.frame.size.width)
+        let width = self.view.frame.size.width
+        let constant = max(width - rect.size.width - leftRightPadding, CGFloat.min)
+        if chat.uid == self.user?.uid ?? "" {
+          let colors = ChatCollectionViewCell.selfColors
+          cell.containerView.backgroundColor = colors.background
+          cell.textLabel.textColor = colors.text
+          cell.trailingConstraint.active = false
+          cell.leadingConstraint.constant = constant
+          cell.leadingConstraint.active = true
+        } else {
+          let colors = ChatCollectionViewCell.othersColors
+          cell.containerView.backgroundColor = colors.background
+          cell.textLabel.textColor = colors.text
+          cell.leadingConstraint.active = false
+          cell.trailingConstraint.constant = constant
+          cell.trailingConstraint.active = true
+        }
+      }
+      
+      // Unfortunately neither FirebaseArray nor FirebaseCollectionViewDataSource
+      // provide an event listener, so in order to scroll on new insertions we
+      // still need to use the query directly.
+      self.query!.observeEventType(.ChildAdded, withBlock: { [unowned self] _ in
+        self.scrollToBottom(animated: true)
+      })
+    }
+    
+    // Notification boilerplate to handle keyboard appearance/disappearance
+    NSNotificationCenter.defaultCenter().addObserver(self,
+                                                     selector: #selector(keyboardWillShow),
+                                                     name: UIKeyboardWillShowNotification,
+                                                     object: nil)
+    NSNotificationCenter.defaultCenter().addObserver(self,
+                                                     selector: #selector(keyboardWillHide),
+                                                     name: UIKeyboardWillHideNotification,
+                                                     object: nil)
+  }
+  
+  override func viewWillDisappear(animated: Bool) {
+    super.viewWillDisappear(animated)
+    NSNotificationCenter.defaultCenter().removeObserver(self)
+  }
+  
+  deinit {
+    NSNotificationCenter.defaultCenter().removeObserver(self)
+  }
+  
+  @objc private func keyboardWillShow(notification: NSNotification) {
```

Comment:
This is boilerplatey. For sample code where we expect code readers to be higher than normal its worth organizing the code so the demonstrative stuff is higher up, and the generic stuff lower down (within bounds of normal language conventions).